### DISCUSSION TITLE
Enable PKCE for OIDC flows

### DISCUSSION THREAD

**Discussion by micolous** [outline/outline#7706] [Ideas]
Outline's OIDC authentication provider currently *always* disables [PKCE](https://www.rfc-editor.org/rfc/rfc7636.html) support:

https://github.com/outline/outline/blob/4f84daf558615b5b7241966a2e5b729755038339/plugins/oidc/server/auth/oidc.ts#L63

This appears to be from [the original PR for generic OAuth2 support](https://github.com/outline/outline/pull/2388).

At present, Outline **cannot** authenticate to a server which **requires** PKCE support.

## Background

Per [RFC 7636 section 5](https://www.rfc-editor.org/rfc/rfc7636.html#section-5):

> Server implementations of this specification MAY accept OAuth2.0
> clients that do not implement this extension.  If the "code_verifier"
> is not received from the client in the Authorization Request, servers
> supporting backwards compatibility revert to the OAuth 2.0 [[RFC6749](https://www.rfc-editor.org/rfc/rfc6749)]
> protocol without this extension.

"MAY" indicates that the fallback mechanism is truly optional; so an authorisation provider could *require* *all* clients support PKCE for security reasons (regardless of whether the security weakness outlined in RFC 7636 applies to them).

Edit: more background reading: https://kanidm.github.io/kanidm/master/frequently_asked_questions.html#why-is-disabling-pkce-considered-insecure

**Comment by tommoor** [comment]
That's right PKCE is applicable for embedded "public" client apps that cannot secure a secret. Outline's OAuth implementation is server side so there is no need for this.

**Reply by micolous** [reply]
There's nothing stopping a hosted web app from implementing PKCE, especially given it seems to be a boolean away. 😉

**Reply by FilipeRamalho** [reply]
The draft document from IETF outlining OAuth 2.0 Security Best Current Practice says:
> For confidential clients, the use of PKCE [RFC7636] is RECOMMENDED, as it provides strong protection against misuse and injection of authorization codes as described in Section 4.5.3.1 and, as a side-effect, prevents CSRF even in the presence of strong attackers as described in Section 4.7.1.

https://datatracker.ietf.org/doc/draft-ietf-oauth-security-topics/

The draft for OAuth 2.1 even mandates the use of PKCE.
https://datatracker.ietf.org/doc/draft-ietf-oauth-v2-1/

On OAuth.net, they say
> PKCE is recommended even if a client is using a client secret or other form of [client authentication](https://oauth.net/2/client-authentication/) like [private_key_jwt](https://oauth.net/private-key-jwt).
PKCE was originally designed to protect the authorization code flow in mobile apps, but its ability to prevent authorization code injection makes it useful for every type of OAuth client, even web apps that use client authentication.

https://oauth.net/2/pkce/

> It is recommended that all clients use the [PKCE](https://oauth.net/2/pkce/) extension with this flow as well to provide better security.

https://oauth.net/2/grant-types/authorization-code/

OAuth.net is run by Aaron Parecki, which is part of the OAuth Working Group at the IETF, Director at Okta.

Using PKCE is best practice, while it only mitigates sophisticated attacks in the case of confidential clients, it's also so easy to implement, especially in this case. I really would recommend at the very least explicitly giving the option and adding that to the docs. The very best thing would to switch it to default on new installations, the second best thing to give the option and advertise it.

**Comment by pelag0s** [comment]
We now have to disable `PKCE_REQUIRED` in our django-oauth-tookit deployment because of Outline, which also decreases the security of all other clients.

Would really appreciate this being added to Outline.

**Reply by tommoor** [reply]
No idea why this is disabled to be honest, we will enable and test that it continues to work for existing installs

**Reply by tommoor** [reply]
qq does your `well-known` url include `code_challenge_methods_supported` like Google's here?

https://accounts.google.com/.well-known/openid-configuration

**Reply by DerekWolfie** [reply]
Any updates on that matter?
It still doesn't work for me

**Reply by tommoor** [reply]
It hasn't been touched, I can take a look again

**Reply by tommoor** [reply]
Nevermind, I was looking at the old file - it is enabled now if OIDC server advertises support:

https://github.com/outline/outline/blob/main/plugins/oidc/server/auth/oidc.ts#L54

**Reply by m8426** [reply]
@DerekWolfie make sure you're using the discovery path, if you hardcode the urls, pretty sure PKCE doesn't work

**Comment by DerekWolfie** [comment]
@m8426 and @tommoor managed to get it working.
I had to open the code to findout that if I fillout a few of the env vars it could no longer fall in oidc discovery mode.
Thank you very much for reaching back.
Very appreciated!
